pkg/util/podstore/podstore.go

```diff
@@ -0,0 +1,227 @@
+	s := &PodStore{
+		pods:         cache.NewIndexer(podKeyFunc, cache.Indexers{podIPIndex: podIPIndexFunc}),
+		podsToDelete: workqueue.NewDelayingQueueWithCustomClock(clock, deleteQueueName),
+		clock:        clock,
+		timestampMap: map[types.UID]*podLocalTimestamps{},
+	}
+	podInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    s.onPodCreate,
+			UpdateFunc: s.onPodUpdate,
+			DeleteFunc: s.onPodDelete,
+		})
+	return s
+}
+
+func NewPodStore(podInformer cache.SharedIndexInformer) *PodStore {
+	return NewPodStoreWithClock(podInformer, clock.RealClock{})
+}
+
+func (s *PodStore) onPodUpdate(oldObj interface{}, newObj interface{}) {
+	newPod, ok := newObj.(*corev1.Pod)
+	if !ok {
+		klog.ErrorS(nil, "Received unexpected object", "newObj", newObj)
+		return
+	}
+	s.pods.Update(newPod)
+	klog.V(4).InfoS("Process Pod Update Event", "Pod", klog.KObj(newPod))
+}
+
+func (s *PodStore) onPodCreate(obj interface{}) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	timeNow := s.clock.Now()
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		klog.ErrorS(nil, "Received unexpected object", "obj", obj)
+		return
+	}
+	err := s.pods.Add(pod)
+	if err != nil {
+		klog.ErrorS(err, "Error when adding pod to curPods")
+		return
+	}
+	switch pod.Status.Phase {
+	case corev1.PodPending:
+		s.timestampMap[pod.UID] = &podLocalTimestamps{CreationTimestamp: timeNow}
+	default:
+		s.timestampMap[pod.UID] = &podLocalTimestamps{CreationTimestamp: pod.CreationTimestamp.Time}
+	}
+	klog.V(4).InfoS("Process Pod Create Event", "Pod", klog.KObj(pod))
+}
+
+func (s *PodStore) onPodDelete(obj interface{}) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	timeNow := s.clock.Now()
+	pod, ok := obj.(*corev1.Pod)
+	if !ok {
+		var err error
+		pod, err = s.checkDeletedPod(obj)
+		if err != nil {
+			klog.ErrorS(err, "Got error while processing Delete Event")
+			return
+		}
+	}
+	timestamp, ok := s.timestampMap[pod.UID]
+	if !ok {
+		klog.ErrorS(nil, "Cannot find podLocalTimestamps in timestampMap", "UID", pod.UID)
+		return
+	}
+	timestamp.DeletionTimestamp = &timeNow
+	s.podsToDelete.AddAfter(pod, delayTime)
+	klog.V(4).InfoS("Process Pod Delete Event", "Pod", klog.KObj(pod))
+}
+
+func (s *PodStore) checkDeletedPod(obj interface{}) (*corev1.Pod, error) {
+	deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+	if !ok {
+		return nil, fmt.Errorf("received unexpected object: %v", obj)
+	}
+	pod, ok := deletedState.Obj.(*corev1.Pod)
+	if !ok {
+		return nil, fmt.Errorf("DeletedFinalStateUnknown object is not of type Pod: %v", deletedState.Obj)
+	}
+	return pod, nil
+}
+
+func (s *PodStore) GetPodByIPAndTime(ip string, startTime time.Time) (*corev1.Pod, bool) {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+	pods, _ := s.pods.ByIndex(podIPIndex, ip)
+	if len(pods) == 0 {
+		return nil, false
+	} else if len(pods) == 1 {
+		// In case the clocks may be skewed in different nodes, we directly return the Pod if there is only
```

Comment:
s/may be skewed in different nodes/may be skewed between different Nodes in the cluster

Reply:
Fixed.